File: src/main/java/org/nygenome/als/graphdb/model/HumanTissueAtlas.java
package org.nygenome.als.graphdb.model;

import org.apache.commons.csv.CSVRecord;

import javax.annotation.Nonnull;
import lombok.Builder;
import lombok.Data;
import lombok.extern.log4j.Log4j;
import org.nygenome.als.graphdb.service.UniProtMappingService;
import org.nygenome.als.graphdb.util.TsvRecordStreamSupplier;

import java.nio.file.Paths;
import java.util.Optional;
import java.util.function.Function;

@Log4j
@Data
@Builder
public class HumanTissueAtlas extends ModelObject {
  /*
  TSV record header
  Gene	Gene name	Tissue	Cell type	Level	Reliability
   */
  private String ensemblGeneId;
  private String geneName;
  private String tissue;
  private String cellType;
  private String level;
  private String reliability;
  private String ensemblTranscriptId;

  private static String resolveTranscriptId(String geneId) {
    Optional<UniProtMapping> uniOpt = UniProtMappingService.INSTANCE.resolveUniProtMappingByEnsemblGeneId(geneId);
    return (uniOpt.isPresent()) ? uniOpt.get().getEnsemblTranscriptId() :"";
  }

  private static HumanTissueAtlas parseCsvRecord (@Nonnull CSVRecord record){
    return HumanTissueAtlas.builder()
        .ensemblGeneId(record.get("Gene"))
        .geneName(record.get("Gene name"))
        .tissue(record.get("Tissue"))
        .cellType(record.get("Cell type"))
        .level(record.get("Level"))
        .reliability(record.get("Reliability"))
        .ensemblTranscriptId(resolveTranscriptId(record.get("Gene")))
        .build();
  }

  public static Function<CSVRecord,HumanTissueAtlas> parseCsvRecordFunction =
      record -> HumanTissueAtlas.parseCsvRecord(record);

  public static void main(String[] args) {
   new TsvRecordStreamSupplier(Paths.get("/data/als/HumanTissueAtlas.tsv"))
       .get()
       .limit(50)
       .map(HumanTissueAtlas.parseCsvRecordFunction)
       .forEach(System.out::println);
  }

}

File: src/main/scala/org/nygenome/als/graphdb/value/HumanTissueAtlas.scala
package org.nygenome.als.graphdb.value

import java.util.Optional
import org.apache.commons.csv.CSVRecord
import org.nygenome.als.graphdb.service.UniProtMappingService
/*
TSV record header
Gene	Gene name	Tissue	Cell type	Level	Reliability
*/

case class HumanTissueAtlas(ensemblGeneId: String, geneName: String, tissue: String,
                          cellType: String, level: String, reliability: String,
                          ensemblTranscriptId: String, uniprotId: String
                         ) {
var resolveTissueCellTypeLabel: String = tissue + ":" + cellType
var proteinTissueKey: Tuple2[String, String] =
  new Tuple2[String, String](uniprotId, resolveTissueCellTypeLabel)

}


object HumanTissueAtlas extends ValueTrait {

private def resolveUniprotId(geneId: String): String = {
  val uniOpt: Optional[UniProtMapping] = UniProtMappingService.INSTANCE.resolveUniProtMappingByEnsemblGeneId(geneId)
  return if ((uniOpt.isPresent)) {
    uniOpt.get.uniPotId
  }
  else {
    ""
  }
}

private def resolveTranscriptId(geneId: String): String = {
  val uniOpt: Optional[UniProtMapping] = UniProtMappingService.INSTANCE.resolveUniProtMappingByEnsemblGeneId(geneId)
  return if ((uniOpt.isPresent)) {
    uniOpt.get.ensemblGeneId
  }
  else {
    ""
  }
}

def parseCSVRecord(record: CSVRecord): HumanTissueAtlas = {
  HumanTissueAtlas(record.get("Gene"), record.get("Gene name"),
    record.get("Tissue"), record.get("Cell type"),
    record.get("Level"), record.get("Reliability"),
    resolveTranscriptId(record.get("Gene")),
    resolveUniprotId(record.get("Gene"))
  )
}
}
